net/yggdrasil/files/yggdrasil.sh

```diff
@@ -97,6 +106,52 @@ proto_yggdrasil_generate_keypair() {
 	public_key=${PrivateKey:64}
 }
 
+proto_yggdrasil_allocate_listen_addresses() {
+	local config="$1"
+
+	# Collect already defined protocols
+	protocols=""
+	_add_address_protocol() {
+		protocols="${protocols}$(echo $1 | cut -d "://" -f1) "
+	}
+	config_list_foreach "$config" listen_address _add_address_protocol
+
+	# Add new address for each previously unspecified protocol
+	for protocol in "tls" "quic"; do
+		if ! echo "$protocols" | grep "$protocol" &>/dev/null; then
+			# By default linux dynamically alocates ports in the range 32768..60999
+			# `sysctl net.ipv4.ip_local_port_range`
+			random_port=$(( ($RANDOM + $RANDOM) % 22767 + 10000 ))
+			proto_yggdrasil_add_string "${protocol}://127.0.0.1:${random_port}"
+		fi
+	done
+}
+
+proto_yggdrasil_generate_jumper_config() {
+	local config="$1"
+	local ygg_sock="$2"
+	local ygg_cfg="$3"
+
+	# Autofill Yggdrasil listeners
+	config_get is_autofill_listeners "$config" "jumper_autofill_listen_addresses"
+	if [ "$is_autofill_listeners" == "1" ]; then
+		echo "yggdrasil_listen = ["
```

Comment:
We generate a config by echoing to stdout?

Reply:
I think explicit redirection on the caller's end looks more concise then repeatedly doing the same after each `echo` call.

Though I probably should have called it like something that "prints" rather than "generates".